### ISSUE TITLE
Date objects turn into string

### ISSUE THREAD

**Issue by Parziphal** [balderdashy/sails-mongo#325]
Hello. I've found this strange behaviour:

I create a document like this:

``` js
User.create({
  conversations: [
    {
      messages: [
        {
          createdAt: new Date()
        }
      ]
    }
  ]
}).exec(function() {});
```

Using Robomongo, I check the new document and I see that the `conversations.messages.createdAt` field is a Date object, so everything's fine. But then I append a new object to the `messages` array like this:

``` js
User.find({}).limit(1).exec(function(err, users) {
  users[0].conversations[0].messages.push({createdAt: new Date()});
  users[0].save();
});
```

And suddenly, all `conversations.messages.createdAt` fields turn into `string` objects. Is this an intended behaviour? Am I missing something? Maybe this will be fixed when this feature is added https://github.com/balderdashy/sails-mongo/issues/44 ?

**EDIT**

Actually any `save` operation screws the data, turning Date objects into strings.

**Comment by sailsbot** [comment]
Thanks for posting, @Parziphal.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
What version of waterline and the adapter are you using?

**Comment by Parziphal** [comment]
`npm list -g` says:
waterline@0.10.26

**Comment by sailsbot** [comment]
Thanks for posting, @Parziphal.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by obj63mc** [comment]
Just wondering if there is any update on this, major bug for example - 

    Post.findOne({postid:req.query.postid}).exec(function(err, p){
        p.postdate = new Date();
        p.save();
    });

This is always saving the date as a string not an ISOdate in mongo.  Any ideas how to work around it?  I am using sails-mongo 0.11.5 and sails 0.11.3

**Comment by devinivy** [comment]
I don't use this adapter, and I haven't had time to look into it.  Do the default `createdAt` and `updatedAt` fields have this problem?

**Comment by obj63mc** [comment]
The created and updated work as expected, it is only when you are using a schemaless type object and setting a date field somewhere in the object.  This did not used to happen in older versions though.  Not sure which versions but I know for sure it worked in 0.9x

**Comment by particlebanana** [comment]
I tried the examples in this issue with the latest version of the adapter and didn't have any issues. Can someone post a link to a repo that reproduces this issue so I can get to the bottom of it?

Specifically can I see model definitions and a set of queries I can run that shows the error?

**Comment by time2try** [comment]
@particlebanana In reference to your comment I have added the following to demonstrate how our team is facing the same issues. I hope this helps you reproduce this issue.

**MODEL:** Twitter.js
```javascript
module.exports = {
  tableName : 'Twitter'
};
```


**CONTROLLER:** TweetController.js
```javascript
replySent: function(req, res){
    var params = req.params.all();
    var tweetID = params.tweetID;
    var reply = params.reply;

    var replyBlock = {
      date_replied: new Date(),
      content: reply
    };
        Twitter.findOne({'tweet_id': tweetID})
          .exec(function (err, tweet) {
            if (err) {
              return res.serverError(err);
            } else {
              if(_.isEmpty(tweet.reply)){
                tweet.reply= [replyBlock];
              }else{
                tweet.reply.push(replyBlock);
              }
              tweet.save(function(err){
                if(err) return res.serverError(err);
                return res.ok('Success!');
              });
            }
          });
}
```


**Tweet Object** in MongoDB before the `tweet.save(function(err){...})`. Notice the `ISODate()` on `date_published`.
```bson
{ 
    "_id" : ObjectId("5762d019d432601501bfe083"), 
    "tweet_id" : "743475907659079680", 
    "retweet_count" : 0, 
    "content" : "Wow! Kony, Inc. is ahead of the pack! Most Visionary Leader-Gartner MQ for Mobile Application Development Platforms! <a href=https://t.co/1QvGHxV9k5>https://t.co/1QvGHxV9k5</a> ", 
    "date_published" : ISODate("2016-06-16T16:10:44.000+0000"), 
    "username" : "jeankondo1", 
    "extracted_urls" : [
        "https://t.co/1QvGHxV9k5"
    ], 
    "search_term" : "\"application development\" -job", 
    "analyzed_location" : {
        "province" : "US Southwest", 
        "source" : "Author"
    }
}
```

**Tweet Object** after `tweet.save(function(err){...})`. Notice `date_published` now. The `ISODate()` has been removed and replaced by a `String`.
```bson
{ 
    "_id" : ObjectId("5762d04cd06c833168bd9851"), 
    "tweet_id" : "743475907659079680", 
    "retweet_count" : NumberInt(0), 
    "content" : "Wow! Kony, Inc. is ahead of the pack! Most Visionary Leader-Gartner MQ for Mobile Application Development Platforms! <a href=https://t.co/1QvGHxV9k5>https://t.co/1QvGHxV9k5</a> ", 
    "date_published" : "2016-06-16T16:10:44.000Z", 
    "username" : "jeankondo1", 
    "extracted_urls" : [
        "https://t.co/1QvGHxV9k5"
    ], 
    "search_term" : "\"application development\" -job",  
    "analyzed_location" : {
        "province" : "US Southwest", 
        "source" : "Author"
    }, 
    "reply" : [
        {
            "date_replied" : "2016-06-16T17:46:54.951Z", 
            "content" : "hello mate"
        }
    ], 
    "updatedAt" : ISODate("2016-06-16T17:46:55.029+0000")
}
```

**Comment by sailsbot** [comment]
This issue was moved to balderdashy/sails#4020
